File: Blind 75/01 - Arrays and Hashing/49-group-anagrams.py
'''
Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
'''
import collections
from time import time
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        buckets = {}
        for word in strs:
            # Initialize map
            a2z_hash_key = [0] * 26
            for l in word:
                a2z_hash_key[ord(l) - ord('a')] += 1

            a2z_hash_key = str(a2z_hash_key)
            if a2z_hash_key not in buckets:
                buckets[a2z_hash_key] = [word]
            else:
                buckets[a2z_hash_key].append(word)

        return [group for group in buckets.values()]

    def reference(self, strs: List[str]) -> List[List[str]]:
        ans = collections.defaultdict(list)

        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            ans[tuple(count)].append(s)
        return ans.values()

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.groupAnagrams(case))
                else:
                    self.groupAnagrams(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        ['eat', 'tea', 'tan', 'ate', 'nat', 'bat'],
        [''],
        ['a'],
    ]
    test.quantify(test_cases)

File: Blind 75/01 - Arrays and Hashing/JavaScript/49-group-anagrams.js
const { performance } = require('perf_hooks');

const CODES = {
  a: 0,
  b: 1,
  c: 2,
  d: 3,
  e: 4,
  f: 5,
  g: 6,
  h: 7,
  i: 8,
  j: 9,
  k: 10,
  l: 11,
  m: 12,
  n: 13,
  o: 14,
  p: 15,
  q: 16,
  r: 17,
  s: 18,
  t: 19,
  u: 20,
  v: 21,
  w: 22,
  x: 23,
  y: 24,
  z: 25,
};

class Solution {
  method(strs) {
    let anagrams = {};
    for (const w of strs) {
      let id = new Array(26).fill(0);
      for (const c of w) {
        id[c.charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
      }
      id = id.join('');

      if (id in anagrams) {
        anagrams[id].push(w);
      } else {
        anagrams[id] = [w];
      }
    }

    return Object.values(anagrams);
  }

  reference(strs) {
    const map = Object.create(null);
    for (const word of strs) {
      const hash = this.hashWord(word);
      if (!(hash in map)) {
        map[hash] = [];
      }
      map[hash].push(word);
    }

    const groups = [];
    for (const key in map) {
      groups.push(map[key]);
    }
    return groups;
  }

  hashWord = (word) => {
    const hash = new Array(26).fill(0);
    for (const ch of word) {
      ++hash[CODES[ch]];
    }
    return hash.toString();
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}
const test = new Solution();
const testCases = [['eat', 'tea', 'tan', 'ate', 'nat', 'bat'], [''], ['a']];
test.quantify(testCases);
